File: GhostBot/functions/runner.py
from __future__ import annotations

import logging
import time
from abc import ABC, abstractmethod
from typing import TYPE_CHECKING, Callable, Any

from GhostBot import logger as _logger
from GhostBot.enums.bot_status import BotStatus
from GhostBot.lib.math import linear_distance

if TYPE_CHECKING:
    from GhostBot.controller.bot_controller import BotClientWindow


def run_at_interval(run_on_start: bool = False, run_in_battle: bool = False):
    def inner(_clazz):
        _init = _clazz.__init__
        def init(self, *args, **kwargs):
            self._last_time_ran = 0 if run_on_start else time.time()
            ret = _init(self, *args, **kwargs)

            if hasattr(_clazz, '_setup'):
                self._log_debug("%s :: %s :: running function setup", self.__class__.__name__, self._client.name)
                _clazz._setup(self)

            if not hasattr(self, '_interval'):
                raise AttributeError(f"Abstract property _interval not defined for {self.__class__.__name__}")
            return ret

        _run = _clazz.run
        def run(self: Runner, *args, **kwargs):
            if should_run(self):
                self._log_debug("%s :: %s :: running function", self.__class__.__name__, self._client.name)
                self._last_time_ran = time.time()
                _run(self, *args, **kwargs)

        def should_run(self):
            if not run_in_battle and self._client.in_battle:
                return False
            return time.time() - self._last_time_ran > self._interval

        _clazz.__init__ = init
        _clazz.run = run
        return _clazz
    return inner

# TODO: class that interacts with NPC's, probably extend Locational

class InjectedLoggingMixin(ABC):

    def __init__(self, client: BotClientWindow):
        self.logger = _logger.getChild(self.__class__.__name__)
        self._error_loggers = [self.logger.error]
        self._info_loggers = [self.logger.info]
        self._debug_loggers = [self.logger.debug]
        self._client = client
        if not self.__class__.__name__.endswith('Context'):
            self._log_debug(f"initializing {self.__class__.__name__}...")

    def add_logger(self, _logger: Callable[[str], Any], level: int = logging.INFO):
        if level < logging.INFO:
            self._debug_loggers.append(_logger)
        if level < logging.ERROR:
            self._info_loggers.append(_logger)
        self._error_loggers.append(_logger)

    def _log_err(self, msg: str, *args) -> None:
        all(map(lambda f: f(f"{self._client.name}: {msg}", *args), self._error_loggers))

    def _log_info(self, msg: str, *args) -> None:
        all(map(lambda f: f(f"{self._client.name}: {msg}", *args), self._info_loggers))

    def _log_debug(self, msg: str, *args) -> None:
        all(map(lambda f: f(f"{self._client.name}: {msg}", *args), self._debug_loggers))

class Runner(InjectedLoggingMixin, ABC):
    """
    base class for any optional function to be run on the bot, eg. fairy, attack, ...
    """
    def __init__(self, client: BotClientWindow):
        super().__init__(client)

    def run(self):
        if self._client.bot_status == BotStatus.running:
            return self._run()
        self._log_debug("not running as client not in running status.")
        return None

    @abstractmethod
    def _run(self) -> bool: ...


class Locational(Runner, ABC):
    """
    Represents a function that has a concept of location.
    """

    def __init__(self, client: BotClientWindow):
        super().__init__(client)
        self.start_location: tuple[int, int] = self.determine_start_location()

    def determine_start_location(self):
        """Returns either the config stored attack_spot, or the current location of the char as the `start_location`"""
        if (attack := self._client.config.attack) is not None:
            if attack.spot is not None:
                return int(attack.spot[0]), int(attack.spot[1])
        return self._client.location

    def _goto_start_location(self):
        """Moves the char to the saved `start_location`"""
        while linear_distance(self.start_location, self._client.location) > 2 and self._client.running:
            self._log_debug(f'{self._client.name}: go to saved spot: {self.start_location}')
            self._client.move_to_pos(self.start_location)

File: kotlin-server/src/main/kotlin/com/ghostbot/functions/Runner.kt
package com.ghostbot.functions

import com.ghostbot.controller.BotClientWindow
import com.ghostbot.lib.linearDistance
import com.ghostbot.lib.sleepSeconds
import java.util.logging.Level

/**
 * Port of GhostBot/functions/runner.py.
 *
 * Python's `@run_at_interval` metaprogramming is expressed as three overridable
 * hooks on [Runner]: [isIntervalFunction], [intervalGate] and the
 * [lastTimeRan] bookkeeping, driven from [run].
 */

/**
 * Base mixin providing client-scoped logging, mirroring `InjectedLoggingMixin`.
 */
abstract class InjectedLoggingMixin(val client: BotClientWindow) {
    val logger = java.util.logging.Logger.getLogger("GhostBot.Runner")
    private val errorLoggers = mutableListOf<(String) -> Unit>()
    private val infoLoggers = mutableListOf<(String) -> Unit>()
    private val debugLoggers = mutableListOf<(String) -> Unit>()

    init {
        if (!javaClass.simpleName.endsWith("Context")) {
            _logDebug("initializing ${javaClass.simpleName}...")
        }
    }

    /** Port of `add_logger` — register an extra consumer for a level. */
    fun addLogger(logger: (String) -> Unit, level: Level = Level.INFO) {
        if (level.intValue() < Level.INFO.intValue()) debugLoggers.add(logger)
        if (level.intValue() < Level.SEVERE.intValue()) infoLoggers.add(logger)
        errorLoggers.add(logger)
    }

    protected fun _logErr(msg: String) {
        val m = "${client.name}: $msg"
        logger.log(Level.SEVERE, m)
        errorLoggers.forEach { it(m) }
    }

    protected fun _logInfo(msg: String) {
        val m = "${client.name}: $msg"
        logger.log(Level.INFO, m)
        infoLoggers.forEach { it(m) }
    }

    protected fun _logDebug(msg: String) {
        val m = "${client.name}: $msg"
        logger.log(Level.FINE, m)
        debugLoggers.forEach { it(m) }
    }
}

/**
 * Base class for any optional function to be run on the bot (attack, regen, ...).
 */
abstract class Runner(client: BotClientWindow) :
    InjectedLoggingMixin(client) {

    /** Port of the `@run_at_interval(run_on_start, run_in_battle)` bookkeeping. */
    protected var lastTimeRan: Long = if (runOnStart) 0 else System.currentTimeMillis()
    protected open val runOnStart: Boolean get() = false
    protected open val runInBattle: Boolean get() = false
    /** Interval in ms; null for non-interval functions (run every tick). */
    protected open val intervalMs: Long? get() = null

    /** Non-interval functions: always eligible. Interval functions: gate on battle/interval. */
    protected open fun intervalGate(): Boolean {
        val interval = intervalMs
        if (interval == null) return true
        if (!runInBattle && client.inBattle) return false
        return System.currentTimeMillis() - lastTimeRan > interval
    }

    fun run() {
        if (client.botStatus != com.ghostbot.enums.BotStatus.RUNNING) {
            _logDebug("not running as client not in running status.")
            return
        }
        if (!intervalGate()) return
        val interval = intervalMs
        if (interval != null) lastTimeRan = System.currentTimeMillis()
        _run()
    }

    /** Port of `_run` — returns True when the function did meaningful work. */
    protected abstract fun _run(): Boolean?
}

/**
 * A function that has a concept of a home/start location.
 */
abstract class Locational(client: BotClientWindow) :
    Runner(client) {

    val startLocation: Pair<Int, Int> = determineStartLocation()

    /** Returns the config attack/fairy spot, or the char's current location. */
    protected open fun determineStartLocation(): Pair<Int, Int> {
        val attack = client.config?.attack
        attack?.let {
            it.spot?.let { s -> return s as Pair<Int, Int> }
        }
        val fairy = client.config?.fairy
        fairy?.let {
            it.spot?.let { s -> return s as Pair<Int, Int> }
        }
        return client.location.x to client.location.y
    }

    /** Moves the char to the saved start location. */
    protected fun gotoStartLocation() {
        while (linearDistance(startLocation, client.location.x to client.location.y) > 2 && client.running) {
            _logDebug("${client.name}: go to saved spot: $startLocation")
            client.moveToPos(startLocation)
        }
    }
}
